File: go/internal/e2e/per_session_auth_test.go
package e2e

import (
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestPerSessionAuth(t *testing.T) {
	ctx := testharness.NewTestContext(t)

	// Create client with COPILOT_DEBUG_GITHUB_API_URL redirected to the proxy
	// so per-session auth token resolution (fetchCopilotUser) is intercepted.
	client := ctx.NewClient(func(opts *copilot.ClientOptions) {
		opts.Env = append(opts.Env, "COPILOT_DEBUG_GITHUB_API_URL="+ctx.ProxyURL)
	})
	t.Cleanup(func() { client.ForceStop() })
	// Register per-token user configs on the proxy
	if err := ctx.SetCopilotUserByToken("token-alice", map[string]interface{}{
		"login":                 "alice",
		"copilot_plan":          "individual_pro",
		"endpoints":             map[string]interface{}{"api": ctx.ProxyURL, "telemetry": "https://localhost:1/telemetry"},
		"analytics_tracking_id": "alice-tracking-id",
	}); err != nil {
		t.Fatalf("Failed to set copilot user for alice: %v", err)
	}

	if err := ctx.SetCopilotUserByToken("token-bob", map[string]interface{}{
		"login":                 "bob",
		"copilot_plan":          "business",
		"endpoints":             map[string]interface{}{"api": ctx.ProxyURL, "telemetry": "https://localhost:1/telemetry"},
		"analytics_tracking_id": "bob-tracking-id",
	}); err != nil {
		t.Fatalf("Failed to set copilot user for bob: %v", err)
	}

	t.Run("should authenticate with per-session token", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			GitHubToken:         "token-alice",
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		authStatus, err := session.RPC.Auth.GetStatus(t.Context())
		if err != nil {
			t.Fatalf("Failed to get auth status: %v", err)
		}

		if !authStatus.IsAuthenticated {
			t.Errorf("Expected session to be authenticated")
		}
		if authStatus.Login == nil || *authStatus.Login != "alice" {
			t.Errorf("Expected login to be 'alice', got %v", authStatus.Login)
		}
	})

	t.Run("should isolate auth between sessions", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		sessionA, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			GitHubToken:         "token-alice",
		})
		if err != nil {
			t.Fatalf("Failed to create session A: %v", err)
		}

		sessionB, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			GitHubToken:         "token-bob",
		})
		if err != nil {
			t.Fatalf("Failed to create session B: %v", err)
		}

		statusA, err := sessionA.RPC.Auth.GetStatus(t.Context())
		if err != nil {
			t.Fatalf("Failed to get auth status for session A: %v", err)
		}

		statusB, err := sessionB.RPC.Auth.GetStatus(t.Context())
		if err != nil {
			t.Fatalf("Failed to get auth status for session B: %v", err)
		}

		if statusA.Login == nil || *statusA.Login != "alice" {
			t.Errorf("Expected session A login to be 'alice', got %v", statusA.Login)
		}
		if statusB.Login == nil || *statusB.Login != "bob" {
			t.Errorf("Expected session B login to be 'bob', got %v", statusB.Login)
		}
	})

	t.Run("should be unauthenticated without token", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		authStatus, err := session.RPC.Auth.GetStatus(t.Context())
		if err != nil {
			t.Fatalf("Failed to get auth status: %v", err)
		}

		// Without a per-session token, there is no per-session identity.
		// In CI the process-level fake token may still authenticate globally,
		// so we check Login rather than IsAuthenticated.
		if authStatus.Login != nil && *authStatus.Login != "" {
			t.Errorf("Expected no per-session login without token, got %q", *authStatus.Login)
		}
	})

	t.Run("should fail with invalid token", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		_, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			GitHubToken:         "invalid-token",
		})
		if err == nil {
			t.Fatal("Expected session creation to fail with invalid token")
		}
		t.Logf("Got expected error: %v", err)
	})
}

File: nodejs/test/e2e/per_session_auth.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it } from "vitest";
import { approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext.js";

describe("Per-session GitHub auth", async () => {
    const { copilotClient: client, openAiEndpoint, env } = await createSdkTestContext();

    // Redirect GitHub API calls (e.g., fetchCopilotUser) to the proxy
    // so per-session auth token resolution can be tested
    env.COPILOT_DEBUG_GITHUB_API_URL = env.COPILOT_API_URL;

    // Configure per-token responses on the proxy.
    // endpoints.api points back to the proxy so subsequent CAPI calls are also intercepted.
    const proxyUrl = env.COPILOT_API_URL;
    await openAiEndpoint.setCopilotUserByToken("token-alice", {
        login: "alice",
        copilot_plan: "individual_pro",
        endpoints: {
            api: proxyUrl,
            telemetry: "https://localhost:1/telemetry",
        },
        analytics_tracking_id: "alice-tracking-id",
    });

    await openAiEndpoint.setCopilotUserByToken("token-bob", {
        login: "bob",
        copilot_plan: "business",
        endpoints: {
            api: proxyUrl,
            telemetry: "https://localhost:1/telemetry",
        },
        analytics_tracking_id: "bob-tracking-id",
    });

    it("should create session with gitHubToken and check auth status", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
            gitHubToken: "token-alice",
        });

        const authStatus = await session.rpc.auth.getStatus();
        expect(authStatus.isAuthenticated).toBe(true);
        expect(authStatus.login).toBe("alice");
        expect(authStatus.copilotPlan).toBe("individual_pro");

        await session.disconnect();
    });

    it("should isolate auth between sessions with different tokens", async () => {
        const sessionA = await client.createSession({
            onPermissionRequest: approveAll,
            gitHubToken: "token-alice",
        });
        const sessionB = await client.createSession({
            onPermissionRequest: approveAll,
            gitHubToken: "token-bob",
        });

        const statusA = await sessionA.rpc.auth.getStatus();
        const statusB = await sessionB.rpc.auth.getStatus();

        expect(statusA.isAuthenticated).toBe(true);
        expect(statusA.login).toBe("alice");
        expect(statusA.copilotPlan).toBe("individual_pro");

        expect(statusB.isAuthenticated).toBe(true);
        expect(statusB.login).toBe("bob");
        expect(statusB.copilotPlan).toBe("business");

        await sessionA.disconnect();
        await sessionB.disconnect();
    });

    it("should return unauthenticated when no token is provided", async () => {
        const session = await client.createSession({
            onPermissionRequest: approveAll,
        });

        const authStatus = await session.rpc.auth.getStatus();
        // Without a per-session GitHub token, there is no per-session identity.
        // In CI the process-level fake token may still authenticate globally,
        // so we check login rather than isAuthenticated.
        expect(authStatus.login).toBeFalsy();

        await session.disconnect();
    });

    it("should error when creating session with invalid token", async () => {
        await expect(
            client.createSession({
                onPermissionRequest: approveAll,
                gitHubToken: "invalid-token-12345",
            })
        ).rejects.toThrow();
    });
});
